Jellyfin.Api/Controllers/SyncPlayController.cs

```diff
@@ -85,38 +87,124 @@ public ActionResult SyncPlayJoinGroup([FromQuery, Required] Guid groupId)
+        /// <param name="requestData">The new playlist to play in the group.</param>
         /// <response code="204">Play request sent to all group members.</response>
         /// <returns>A <see cref="NoContentResult"/> indicating success.</returns>
         [HttpPost("Play")]
         [ProducesResponseType(StatusCodes.Status204NoContent)]
-        public ActionResult SyncPlayPlay()
+        public ActionResult SyncPlayPlay(
+            [FromBody, Required] PlayRequestBody requestData)
+        {
+            var currentSession = RequestHelpers.GetSession(_sessionManager, _authorizationContext, Request);
+            var syncPlayRequest = new PlayGroupRequest(
+                requestData.PlayingQueue,
+                requestData.PlayingItemPosition,
+                requestData.StartPositionTicks);
+            _syncPlayManager.HandleRequest(currentSession, syncPlayRequest, CancellationToken.None);
+            return NoContent();
+        }
+
+        /// <summary>
+        /// Request to change playlist item in SyncPlay group.
+        /// </summary>
+        /// <param name="requestData">The new item to play.</param>
+        /// <response code="204">Queue update request sent to all group members.</response>
+        /// <returns>A <see cref="NoContentResult"/> indicating success.</returns>
+        [HttpPost("SetPlaylistItem")]
+        [ProducesResponseType(StatusCodes.Status204NoContent)]
+        public ActionResult SyncPlaySetPlaylistItem(
+            [FromBody, Required] SetPlaylistItemRequestBody requestData)
         {
             var currentSession = RequestHelpers.GetSession(_sessionManager, _authorizationContext, Request);
-            var syncPlayRequest = new PlaybackRequest()
-            {
-                Type = PlaybackRequestType.Play
-            };
+            var syncPlayRequest = new SetPlaylistItemGroupRequest(requestData.PlaylistItemId);
+            _syncPlayManager.HandleRequest(currentSession, syncPlayRequest, CancellationToken.None);
+            return NoContent();
+        }
+
+        /// <summary>
+        /// Request to remove items from the playlist in SyncPlay group.
+        /// </summary>
+        /// <param name="requestData">The items to remove.</param>
+        /// <response code="204">Queue update request sent to all group members.</response>
+        /// <returns>A <see cref="NoContentResult"/> indicating success.</returns>
+        [HttpPost("RemoveFromPlaylist")]
+        [ProducesResponseType(StatusCodes.Status204NoContent)]
+        public ActionResult SyncPlayRemoveFromPlaylist(
+            [FromBody, Required] RemoveFromPlaylistRequestBody requestData)
+        {
+            var currentSession = RequestHelpers.GetSession(_sessionManager, _authorizationContext, Request);
+            var syncPlayRequest = new RemoveFromPlaylistGroupRequest(requestData.PlaylistItemIds);
+            _syncPlayManager.HandleRequest(currentSession, syncPlayRequest, CancellationToken.None);
```

Comment:
Can't find the old comment in which I've explained a bit why these exist, but briefly, I tried to apply the [Visitor pattern](https://en.wikipedia.org/wiki/Visitor_pattern) along with the [State pattern](https://en.wikipedia.org/wiki/State_pattern) to create what basically is a state machine that handles requests.

The `*RequestBody` DTOs are made for accepting data from the clients while the `*GroupRequest` classes exist for the state machine.

Also, I wanted to make `SyncPlayManager` agnostic to the type of request it handles: its job is to pass these requests along to the correct `GroupController`.

All of these reasons led to this result I guess.

Should I drop the overloading with `HandleRequest` and explicit every request type in `SyncPlayManager`? Like `SyncPlayManager.PlayRequest(RequestDTO)` and so on?

Reply:
I'm not necessarily against either pattern or the relatively thin SyncPlayManager, _but_ it adds a bit of "cruft" and the flow is more complex. For state machines I would probably have opted to use something like https://github.com/dotnet-state-machine/stateless.

For the `HandleRequest` overloads I think that renaming them so that their intent is more visible would go a long way to improve readability. It adds a lot of latency having to look at the request type and then from that infer what the intent is.

Reply:
I've moved the request arguments so they're first, not much but it does help with readability.

I gave a quick look at *Stateless* and seems promising, like it should be able to handle the states I've defined for SyncPlay. Tomorrow I'll see if I can migrate the code to it.